source/event_stream_channel_handler.c

```diff
@@ -0,0 +1,383 @@
+            size_t to_copy = message_cursor.len > remaining_prelude ? remaining_prelude : message_cursor.len;
+
+            if (!aws_byte_buf_write(&event_stream_handler->message_buf, message_cursor.ptr, to_copy)) {
+                event_stream_handler->on_message_received(NULL, aws_last_error(), event_stream_handler->user_data);
+                aws_channel_shutdown(slot->channel, aws_last_error());
+                goto finished;
+            }
+
+            aws_byte_cursor_advance(&message_cursor, to_copy);
+        }
+
+        /* we need to get the prelude so we can get the message length to know how much to read and also
+         * to check the prelude CRC to protect against bit-flips causing us to read to much memory */
+        if (event_stream_handler->message_buf.len == AWS_EVENT_STREAM_PRELUDE_LENGTH) {
+            struct aws_byte_cursor prelude_cursor = aws_byte_cursor_from_buf(&event_stream_handler->message_buf);
+
+            event_stream_handler->running_crc =
+                aws_checksums_crc32(prelude_cursor.ptr, sizeof(uint32_t) + sizeof(uint32_t), 0);
+            aws_byte_cursor_read_be32(&prelude_cursor, &event_stream_handler->current_message_len);
+
+            if (event_stream_handler->current_message_len > AWS_EVENT_STREAM_MAX_MESSAGE_SIZE) {
+                aws_raise_error(AWS_ERROR_EVENT_STREAM_MESSAGE_FIELD_SIZE_EXCEEDED);
+                event_stream_handler->on_message_received(NULL, aws_last_error(), event_stream_handler->user_data);
+                aws_channel_shutdown(slot->channel, aws_last_error());
+                goto finished;
+            }
+
+            /* advance past the headers field since we don't really care about it at this point */
+            aws_byte_cursor_advance(&prelude_cursor, sizeof(uint32_t));
+
+            uint32_t prelude_crc = 0;
+            aws_byte_cursor_read_be32(&prelude_cursor, &prelude_crc);
+
+            /* make sure the checksum matches before processing any further */
+            if (event_stream_handler->running_crc != prelude_crc) {
+                aws_raise_error(AWS_ERROR_EVENT_STREAM_PRELUDE_CHECKSUM_FAILURE);
+                event_stream_handler->on_message_received(NULL, aws_last_error(), event_stream_handler->user_data);
+                aws_channel_shutdown(slot->channel, aws_last_error());
+                goto finished;
+            }
+        }
+
+        /* read whatever is remaining from the message */
+        if (event_stream_handler->message_buf.len < event_stream_handler->current_message_len) {
+            size_t remaining = event_stream_handler->current_message_len - event_stream_handler->message_buf.len;
+            size_t to_copy = message_cursor.len > remaining ? remaining : message_cursor.len;
+
+            struct aws_byte_cursor to_append = aws_byte_cursor_from_array(message_cursor.ptr, to_copy);
+            aws_byte_cursor_advance(&message_cursor, to_copy);
```

Comment:
byte_cursor_advance() returns the portion that was advanced over

struct aws_byte_cursor to_append = aws_byte_cursor_advance(&message_cursor, to_copy);
~~aws_byte_cursor_advance(&message_cursor, to_copy);~~